README.md
# [Power BI](https://powerbi.microsoft.com/) [Python 3](https://www.python.org/) Data Sources for Hornbill Reporting And Trend Engine

## Overview

These example scripts have been provided to enable Power BI administrators to build reports and dashboards using Hornbill Reporting and Advanced Analytics Trend Engine data as their data source(s).

## Dependencies

The scripts have been written in [Python 3](https://www.python.org/), and were developed using the following:

- [Power BI Desktop build 2.75.5649.861 64-bit (November 2019)](https://powerbi.microsoft.com/)
- [Python 3.8](https://docs.python.org/3/whatsnew/3.8.html)

The following packages are required dependencies, and can be installed using the Python Package Installer (pip):

- [requests](https://pypi.org/project/requests/)
- [pandas](https://pypi.org/project/pandas/)
- [xlrd](https://pypi.org/project/xlrd/)

## Configuration used in all scripts

Each script requires the following variables to be set (all case-sensitive):

- `apiKey` - This is an API key generated against a user account on the Hornbill Administration Console, where the user account has sufficient access to run reports and access trending data.
- `instanceId` - This is the (case sensitive) name of the instance to connect to

## Scripts

### PowerBIReport.py

This script will:

- Run a pre-defined report on the Hornbill instance;
- Wait for the report to complete;
- Retrieve the report CSV data and present back as a dataframe called df, which can then be consumed by PowerBI.

Script Variables:

- `reportId`: The ID (Primary Key, INT) of the  report to be run;
- `suspendSeconds`: The number of seconds the script should wait between checks to see if the report is complete;
- `deleteReportInstance`: a boolean value to determine if, once the report is run on Hornbill and the data has been pulled in to PowerBI, whether the historic report run instance should be removed from your Hornbill report;
- `useXLSX`: False = the script will use the CSV output from your report; True = the script will use the XLSX output from your report. NOTE: XLSX output will need to be enabled within the Output Formats > Additional Data Formats section of your report in Hornbill.

### PowerBIHistoricReport.py

This script will:

- Retrieve a historic report CSV from your Hornbill instance;
- Present the report data back as a dataframe called df, which can then be consumed by PowerBI.

Script Variables:

- `reportId`: The ID (Primary Key, INT) of the  report to be run;
- `reportRunId`: The Run ID (INT) of a historic run of the above report ID;
- `useXLSX`: False = the script will use the CSV output from your report; True = the script will use the XLSX output from your report. NOTE: XLSX output will need to be enabled within the Output Formats > Additional Data Formats section of your report in Hornbill.

### PowerBITrendingData.py

This script will:

- Run the reporting::measureGetInfo API  against your Hornbill instance, with a given measure ID (Primary Key, INT);
- Build a table containing all Trend Value entries for the selected measure;
- Present the trend data back as a dataframe called df, which can then be consumed by PowerBI.

Script Variable:

- `measureId`: The ID (Primary Key, INT) of the measure to return trend data from.

Outputs:
As the response parameters from the Trending Engine is fixed (unlike the Reporting engine, which has user-specified column outputs), the output for this report will always consist  of the following columns:

- `value`: the value of the trend sample;
- `sampleId`: the ID of the sample;
- `sampleTime`: the time & date that the sample was taken;
- `dateRange.from`: the start date of the sample snapshot;
- `dateRange.to`: the end date of the sample snapshot;

## Power BI Notes

Please see the [Power BI Documentation](https://docs.microsoft.com/en-us/power-bi/desktop-python-scripts) for more information about using Python with Power BI.

These scripts have been designed to be used as data sources only, and not as the source of Python visuals within Power BI. Which is not to say they couldn't be used in your Python visuals, with a little extra code and the [matplotlib](https://pypi.org/project/matplotlib/) library!


CHANGELOG.md
# CHANGELOG

## 1.1.0 (12th November, 2020)

Features:

- Added support for using the XLSX report output for Report and HistoricReport scripts, rather than the default CSV

## 1.0.0 (2nd January, 2020)

Features:

- Initial Release


PowerBITrendingData.py
# PowerBITrendingData.py
# v1.0.0 02/01/2020
# This script will retrieve trending data from a measure on a Hornbill instance, and output the data in a Power BI consumable dataframe

from pandas.io.json import json_normalize
import requests
import sys
import time
import io
import json

apiKey = 'yourapikey'           # This is the (case sensitive) API key to use to authenticate the API calls against the Hornbill instance
instanceId = 'yourinstanceid'   # This is the (case sensitive) ID of your Hornbill instance
measureId = 11                  # This is the ID of the measure, which can be taken from the measure URL in the admin console

# Get API endpoint
URL = 'https://files.hornbill.com/instances/{instanceId}/zoneinfo'.format(instanceId=instanceId)
try:
    endpoint = requests.get(url = URL).json()["zoneinfo"]["endpoint"]
except requests.exceptions.RequestException as e:
    sys.exit('Unexpected response when attempting to retrieve Hornbill Zone Information: ' + e)

def invokeXmlmc(service, method, params):
    xmlmcEndpoint = endpoint + "xmlmc/{service}/?method={method}" 
    URL = xmlmcEndpoint.format(service=service, method=method)
    headers = {
        'Content-type': 'application/json',
        'Accept': 'application/json',
        'Authorization': 'ESP-APIKEY ' + apiKey
    }
    data = {
        'methodCall':{
            '@service':service,
            '@method':method,
            'params':params
        }
    }
    try:
        response = requests.post(url=URL, json=data, headers=headers)
        if 200 >= response.status_code <= 299:
            return response.json()
        else:
            return {
                '@status': False,
                'state': {
                    'error': 'Unexpected status returned from call to {service}::{method}: {statusCode}'.format(service=service, method=method, statusCode=response.status_code)
                }
            } 
    except requests.exceptions.RequestException as e:
        return {
            '@status': False,
            'state': {
                'error': 'Unexpected response from call to {service}::{method}: {errorString}'.format(service=service, method=method, errorString=e)
            }
        }

measureAPIJson = {
    'measureId': measureId,
    'returnMeasureValue': 'true',
    'returnMeasureTrendData': 'true'
}
measureData = invokeXmlmc('reporting', 'measureGetInfo', measureAPIJson)
df = json_normalize(measureData['params']['trendValue'])
print(df)

PowerBIHistoricReport.py
# PowerBIHistoricReport.py
# v1.0.0 02/01/2020
# This script will retrieve the CSV data from a historic report on a Hornbill instance, and output the data in a Power BI consumable dataframe

import pandas as pd
import requests
import sys
import time
import io
import os

apiKey = 'yourapikey'           # This is the (case sensitive) API key to use to authenticate the API calls against the Hornbill instance
instanceId = 'yourinstanceid'   # This is the (case sensitive) ID of your Hornbill instance
reportId = 6                    # This is the ID of the report, which can be taken from the report URL in the admin console
reportRunId = 181               # This is the ID of the report run instance, which can be taken from the report history in the admin console
useXLSX = False                 # This specifies whether to use the XLSX output from your target Hornbill Report, rather than the default CSV

# Get API endpoint
URL = 'https://files.hornbill.com/instances/{instanceId}/zoneinfo'.format(instanceId=instanceId)
try:
    endpoint = requests.get(url = URL).json()["zoneinfo"]["endpoint"]
except requests.exceptions.RequestException as e:
    sys.exit('Unexpected response when attempting to retrieve Hornbill Zone Information: ' + e)

def invokeXmlmc(service, method, params):
    xmlmcEndpoint = endpoint + "xmlmc/{service}/?method={method}" 
    URL = xmlmcEndpoint.format(service=service, method=method)
    headers = {
        'Content-type': 'application/json',
        'Accept': 'application/json',
        'Authorization': 'ESP-APIKEY ' + apiKey
    }
    data = {
        'methodCall':{
            '@service':service,
            '@method':method,
            'params':params
        }
    }
    try:
        response = requests.post(url=URL, json=data, headers=headers)
        if 200 >= response.status_code <= 299:
            return response.json()
        else:
            return {
                '@status': False,
                'state': {
                    'error': 'Unexpected status returned from call to {service}::{method}: {statusCode}'.format(service=service, method=method, statusCode=response.status_code)
                }
            } 
    except requests.exceptions.RequestException as e:
        return {
            '@status': False,
            'state': {
                'error': 'Unexpected response from call to {service}::{method}: {errorString}'.format(service=service, method=method, errorString=e)
            }
        }

reportRunDetails = invokeXmlmc('reporting', 'reportRunGetStatus', {'runId': reportRunId})

if useXLSX == True:
    xlsURL = endpoint + "dav/reports/{reportId}/{reportLink}"
    for file in reportRunDetails['params']['files']:
        if file['type'] == 'xlsx':
            fileName = file['name']
    URL = xlsURL.format(reportId=reportId,reportLink=fileName)
    getExcel = requests.get(url=URL, headers={'Authorization':'ESP-APIKEY ' + apiKey})
    open(fileName, 'wb').write(getExcel.content)
    df = pd.read_excel(fileName)
    if os.path.exists(fileName):
        os.remove(fileName)
else:
    csvURL = endpoint + "dav/reports/{reportId}/{reportLink}"
    URL = csvURL.format(reportId=reportId,reportLink=reportRunDetails['params']['reportRun']['csvLink'])
    getCSV = requests.get(url=URL, headers={'Authorization':'ESP-APIKEY ' + apiKey})
    df = pd.read_csv(io.StringIO(getCSV.text))

print(df)

PowerBIReport.py
# PowerBIReport.py
# v1.0.0 02/01/2020
# This script will run a report on a Hornbill instance, and retrieve the data in a Power BI consumable dataframe

import pandas as pd
import requests
import sys
import time
import io
import os

apiKey = 'yourapikey'           # This is the (case sensitive) API key to use to authenticate the API calls against the Hornbill instance
instanceId = 'yourinstanceid'   # This is the (case sensitive) ID of your Hornbill instance
reportId = 6                    # This is the ID of the report, which can be taken from the report URL in the admin console
suspendSeconds = 1              # This is the number of seconds to wait inbetween the checking for report run completion 
deleteReportInstance = True     # This defines whether the report run instance is deleted once the report data has been retrieved
useXLSX = False                 # This specifies whether to use the XLSX output from your target Hornbill Report, rather than the default CSV

# Get API endpoint
URL = 'https://files.hornbill.com/instances/{instanceId}/zoneinfo'.format(instanceId=instanceId)
try:
    endpoint = requests.get(url = URL).json()["zoneinfo"]["endpoint"]
except requests.exceptions.RequestException as e:
    sys.exit('Unexpected response when attempting to retrieve Hornbill Zone Information: ' + e)

def invokeXmlmc(service, method, params):
    xmlmcEndpoint = endpoint + "xmlmc/{service}/?method={method}" 
    URL = xmlmcEndpoint.format(service=service, method=method)
    headers = {
        'Content-type': 'application/json',
        'Accept': 'application/json',
        'Authorization': 'ESP-APIKEY ' + apiKey
    }
    data = {
        'methodCall':{
            '@service':service,
            '@method':method,
            'params':params
        }
    }
    try:
        response = requests.post(url=URL, json=data, headers=headers)
        if 200 >= response.status_code <= 299:
            return response.json()
        else:
            return {
                '@status': False,
                'state': {
                    'error': 'Unexpected status returned from call to {service}::{method}: {statusCode}'.format(service=service, method=method, statusCode=response.status_code)
                }
            } 
    except requests.exceptions.RequestException as e:
        return {
            '@status': False,
            'state': {
                'error': 'Unexpected response from call to {service}::{method}: {errorString}'.format(service=service, method=method, errorString=e)
            }
        }
    
# Kick off report
reportJson = {
    'reportId': reportId
}
reportDetails = invokeXmlmc('reporting', 'reportRun', reportJson)
if reportDetails['@status'] != True:
    sys.exit(reportDetails['state']['error'])

reportRunJson = {'runId': reportDetails['params']['runId']}

reportComplete = False
reportSuccess = False

while reportComplete == False:
    time.sleep(suspendSeconds)
    reportRunDetails = invokeXmlmc('reporting', 'reportRunGetStatus', reportRunJson)
    if reportRunDetails['@status'] != True:
        sys.exit(reportRunDetails['state']['error'])

    if reportRunDetails['params']['reportRun']['status'] == 'completed':
        reportComplete = True
        reportSuccess = True
    elif (reportRunDetails['params']['reportRun']['status'] == 'failed') or (reportRunDetails['params']['reportRun']['status'] == 'aborted'):
        reportComplete = True

if reportSuccess == False:
    sys.exit('Report run was not successful: ' + reportRunDetails['statusMessage'])

if useXLSX == True:
    xlsURL = endpoint + "dav/reports/{reportId}/{reportLink}"
    for file in reportRunDetails['params']['files']:
        if file['type'] == 'xlsx':
            fileName = file['name']
    URL = xlsURL.format(reportId=reportId,reportLink=fileName)
    getExcel = requests.get(url=URL, headers={'Authorization':'ESP-APIKEY ' + apiKey})
    open(fileName, 'wb').write(getExcel.content)
    df = pd.read_excel(fileName)
    if os.path.exists(fileName):
        os.remove(fileName)
else:
    csvURL = endpoint + "dav/reports/{reportId}/{reportLink}"
    URL = csvURL.format(reportId=reportId,reportLink=reportRunDetails['params']['reportRun']['csvLink'])
    getCSV = requests.get(url=URL, headers={'Authorization':'ESP-APIKEY ' + apiKey})
    df = pd.read_csv(io.StringIO(getCSV.text))

if deleteReportInstance == True:
    deleteReport = invokeXmlmc('reporting', 'reportRunDelete', reportRunJson)

print(df)